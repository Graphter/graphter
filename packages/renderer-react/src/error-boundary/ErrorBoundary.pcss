.error {
  background-color: #ff00001f;
  border: 3px dotted red;
  padding: 30px;
  margin: 30px 0;
}