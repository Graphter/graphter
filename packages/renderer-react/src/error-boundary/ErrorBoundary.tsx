import React, {ComponentType} from "react";
import s from './ErrorBoundary.pcss';
import { ErrorRendererProps } from "@graphter/core";

interface Props {
  children: any,
  errorRenderer: ComponentType<ErrorRendererProps>
}

interface State {
  err: Error | null,
  errInfo: any | null
}

export default class ErrorBoundary extends React.Component<Props, State> {
  constructor(props: Props){
    super(props);
    this.state = {
      err: null,
      errInfo: null
    };
  }
  static getDerivedStateFromError(err: Error){
    return { err };
  }
  componentDidCatch(err: Error, errInfo: any){
    this.setState({
      err,
      errInfo
    })
  }
  render(){
    const Error = this.props.errorRenderer
    if(this.state.err !== null){
      return <Error err={this.state.err} />
    }

    return this.props.children || null;
  }
}