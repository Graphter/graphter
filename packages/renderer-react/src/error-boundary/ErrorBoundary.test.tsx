import React, {ComponentType} from 'react';
import { act, render, within } from '@testing-library/react';
import { screen } from '@testing-library/dom'
import ErrorBoundary, {ErrorRendererProps} from "./ErrorBoundary";

function MockErrorRenderer({ err }: ErrorRendererProps){
  return <div data-testid='mock-error'>{err.message}</div>;
}

describe(`<ErrorBoundary />`, () => {

  it(`should render children when no error is raised`, () => {
    const { container } = render(
      <ErrorBoundary errorRenderer={MockErrorRenderer}>
        <div>some child</div>
      </ErrorBoundary>
    )
  });
  it(`should display the error with the defined renderer`, () => {
    function ErroringComponent(){
      throw new Error('This is an error');
      return null;
    }
    act(() => {
      render(
        <ErrorBoundary errorRenderer={MockErrorRenderer}>
          <ErroringComponent />
        </ErrorBoundary>
      );
    });
    const error = screen.getByTestId('mock-error');
    expect(error).toMatchSnapshot();
  });
});

export {};