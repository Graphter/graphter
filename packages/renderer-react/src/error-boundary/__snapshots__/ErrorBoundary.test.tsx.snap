// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ErrorBoundary /> should display the error with the defined renderer 1`] = `
<div
  data-testid="mock-error"
>
  This is an error
</div>
`;
